### DISCUSSION TITLE
using jscad together with three.js

### DISCUSSION THREAD

**Discussion by hrgdavor** [jscad/OpenJSCAD.org#821] [Ideas]
using jscad together with [three.js](https://threejs.org/)

links to topics related to jscad+three.js in user group
 - https://openjscad.nodebb.com/topic/235/threejs-integration
 - https://openjscad.nodebb.com/topic/247/three-js-integration-part2-threejscad

three.js reources/ideas:
 - https://threejsfundamentals.org/ - very nice learning res for three.js
 - physics plugin option 1: https://enable3d.io/examples.html
 - physics plugin option 2: https://chandlerprall.github.io/Physijs/

**Comment by koendv** [comment]
...and use three.js StereoCamera to do rendering for anaglyph 3d glasses.

**Reply by hrgdavor** [reply]
@koendv I have a prototype that uses threejs working on official support for it, and hopefuly babylon too. 

http://3d.hrg.hr/jscad/three/threejscad2.html?uri=#model.screw.js

You can also run your own scripts there by drag and drop. (folders not supported yet)

**Reply by tsdexter** [reply]
@hrgdavor how can I get this for use locally with hot reloading?

**Reply by hrgdavor** [reply]
@tsdexter there is another iteration of this. It is a new prototype. It is not documented or anything, but I use it regularly now.

It can run typescript as well as ES module imports, not just require. 
the repo is here https://github.com/hrgdavor/jscadui it is separate because I need to experiment a lot, and the plan is to then bring back useful bits to jscad. 

I tend to be faster in developing this prototype in the periods interested ppl use bits of it. It would be helpful to at least have someone actualy using the parts of the prototype, and later on even better if others could contribute a bit.

`/apps/engine-test` is a demo app I use to test 3 different engines (regl, three,babylon) but also can be configured to use only one engine 
and be used to develop jscad models (I have been using it for my jscad models actively few months now).

If you would like to colaborate you can also contact me on discord jscad channel https://discord.gg/AaqGskur93

**Comment by koendv** [comment]
On Sat, 18 Sep 2021 07:26:47 -0700
Davor Hrg ***@***.***> wrote:

> @koendv I have a prototype that uses threejs working on official
> support for it, and hopefuly babylon too. 
> 
> http://3d.hrg.hr/jscad/three/threejscad2.html?uri=#model.screw.js
> 
> You can also run your own scripts there by drag and drop. (folders
> not supported yet) 
> 

Feels ok,smooth. This is what I've been doing:

github.com/koendv/openscad-raspberrypi

Would be nice to see something similar in jscad.

koen

**Reply by hrgdavor** [reply]
Cool project. Thanks for sharing. 

I will most definetely. Thanks for the cool idea.
Also side by side, as i have lenses for looking side by side on pc screen

**Reply by hrgdavor** [reply]
@koendv now that I thought about it more... I mostly use one color for models on the project, so the color filter is much better choice as I then have the whole screen to use instead of just half in the side by side stereo render.

**Comment by koendv** [comment]
On Sat, 18 Sep 2021 13:59:33 -0700
Davor Hrg ***@***.***> wrote:

> @koendv now that I thought about it more... I mostly use one color
> for models on the project, so the color filter is much better choice
> as I then have the whole screen to use instead of just half in the
> side by side stereo render. 
> 

For applications like 3d printing, where color matters little, anaglyph
is a reasonable choice. 

But three,js StereoCamera offers anaglyph, parallaxbarrier and
side-by-side, so you could offer all three.

koen

**Comment by hrgdavor** [comment]
@koendv I added the option for all 3, but witout an ui, just special URL for now
(loving threejs more and more :) )

http://3d.hrg.hr/jscad/three/threejscad2.html?uri=model.logos.js
![image](https://user-images.githubusercontent.com/2480762/133942949-218a25c3-3bd0-4d19-a156-66827b9cfc90.png)


http://3d.hrg.hr/jscad/three/threejscad2.html?stereo=Anaglyph&uri=model.logos.js
![image](https://user-images.githubusercontent.com/2480762/133942954-4c9ee1f0-e995-43b1-9ae0-d384fafb5810.png)


http://3d.hrg.hr/jscad/three/threejscad2.html?stereo=SideBySide&uri=model.logos.js
![image](https://user-images.githubusercontent.com/2480762/133942955-24c11226-1c40-4475-81b7-eb2b5ffa23eb.png)


http://3d.hrg.hr/jscad/three/threejscad2.html?stereo=ParallaxBarrier&uri=model.logos.js
![image](https://user-images.githubusercontent.com/2480762/133942959-63f5c09b-73aa-49f2-bb54-e9ca0d4d91c3.png)

**Comment by koendv** [comment]
On Sun, 19 Sep 2021 14:05:31 -0700
Davor Hrg ***@***.***> wrote:

> @koendv I added the option for all 3, but witout an ui, just special
> URL for now (loving threejs more and more :) )
> 

I've tested the anaglyph image, and played around with the "size"
parameter, everything fine.
The speed with which this was developed is remarkable. 

koen

**Reply by hrgdavor** [reply]
> The speed with which this was developed is remarkable.

thanks, but this was already all done in three js, adding it to the prototype was like 10 lines of code :) . So kudos to the threejs team for making so many features into it.
